rln/rln.go

```diff
@@ -181,65 +207,60 @@ func (r *RLN) Poseidon(input ...[]byte) (MerkleNode, error) {
 	return result, nil
 }
 
-func (r *RLN) ExtractMetadata(proof RateLimitProof) (ProofMetadata, error) {
-	externalNullifierRes, err := r.Poseidon(proof.Epoch[:], proof.RLNIdentifier[:])
-	if err != nil {
-		return ProofMetadata{}, fmt.Errorf("could not construct the external nullifier: %w", err)
-	}
-
-	return ProofMetadata{
-		Nullifier:         proof.Nullifier,
-		ShareX:            proof.ShareX,
-		ShareY:            proof.ShareY,
-		ExternalNullifier: externalNullifierRes,
-	}, nil
-}
-
 // GenerateProof generates a proof for the RLN given a KeyPair and the index in a merkle tree.
 // The output will containt the proof data and should be parsed as |proof<128>|root<32>|epoch<32>|share_x<32>|share_y<32>|nullifier<32>|
 // integers wrapped in <> indicate value sizes in bytes
-func (r *RLN) GenerateProof(data []byte, key IdentityCredential, index MembershipIndex, epoch Epoch) (*RateLimitProof, error) {
-	input := serialize(key.IDSecretHash, index, epoch, data)
+func (r *RLN) GenerateProof(
+	data []byte,
+	key IdentityCredential,
+	index MembershipIndex,
+	epoch Epoch,
+	messageId uint32) (*RateLimitProof, error) {
+
+	externalNullifierInput, err := r.Poseidon(epoch[:], RLN_IDENTIFIER[:])
+	if err != nil {
+		return nil, fmt.Errorf("could not construct the external nullifier: %w", err)
+	}
+
+	input := serialize(key.IDSecretHash, index, key.UserMessageLimit, messageId, externalNullifierInput, data)
 	proofBytes, err := r.w.GenerateRLNProof(input)
 	if err != nil {
 		return nil, err
 	}
 
-	if len(proofBytes) != 320 {
-		return nil, errors.New("invalid proof generated")
+	if len(proofBytes) != 288 {
+		return nil, fmt.Errorf("invalid proof generated. size: %d expected: 288",
```

Comment:
Since 288 is used both in this and next line, it's probably better to extract to a constant and then the error should be fmt.Errorf("invalid proof generated. size: %d expected: %d", len(proofBytes), yourConstant)

Reply:
right! 5bd2a083d16a5613b5e273776ef69927f2c1f1ff